---
title: "scATAC-seq Analysis"
author: "KeshavPrasadGubbi"
output:
  html_document:
    toc: yes
    toc_depth: 5
    toc_float: yes
    fig_width: 8
    fig_height: 8
    fig_caption: yes
    number_sections: yes
---

# Initialization

```{r}
suppressPackageStartupMessages(library("ArchR"))
suppressPackageStartupMessages(library("tidyverse"))
suppressPackageStartupMessages(library("here"))
suppressPackageStartupMessages(library("pheatmap"))
suppressPackageStartupMessages(library("ComplexHeatmap"))
suppressPackageStartupMessages(library("patchwork"))
suppressPackageStartupMessages(library("writexl"))
set.seed(1234)

# Set the number of threads
addArchRThreads(threads = 8)

here::i_am("scATAC-seqAnalysis.Rmd")
```
## Input data
```{r}
day1 <- "/media/keshavprasad/HornefLab_Data3/ATACSeq/Data/Day12Sample/2_Processed_data/V16_d12_230105/fragments.tsv.gz"
day5 <- "/media/keshavprasad/HornefLab_Data3/ATACSeq/Data/Day5Sample/2_Processed_data/AJ_v12/fragments.tsv.gz"
day10 <- "/media/keshavprasad/HornefLab_Data3/ATACSeq/Data/Day10Sample/compressed_tars/2_Processed_data/AJ_V12_d10_L129_UI/fragments.tsv.gz"
day25 <- "/media/keshavprasad/HornefLab_Data3/ATACSeq/Data/Day25Sample/fragments.tsv.gz"
day4dpi <- "/media/keshavprasad/HornefLab_Data3/ATACSeq/Data/4dpiSample/fragments.tsv.gz"

# Samples Input Vector
# samples <- c(day1, day5, day10, day25, day4dpi)
# # Set understandable names
# names(samples) <- c("d01", "d05", "d10", "d25", "Inf-d4")
# # inputFiles <- c('scATAC' = samples)
# inputFiles <- c(samples)
# print(names(inputFiles))

# Add respective Genome
addArchRGenome("mm10")
```
```{r}
names(day1) <- c("d01")
inputFile1 <- c('scATAC' = day1)
ArrowFiles1 <- createArrowFiles(
  inputFiles = inputFile1,
  sampleNames = names(day1),
  # filterFrags = 2500,
  minTSS = 8, # 4 was the default value. Dont set this too high;can always increase later
  minFrags = 5500,
  addTileMat = TRUE,
  addGeneScoreMat = TRUE,
  subThreading = TRUE,
  cleanTmp = TRUE,
  force = TRUE, # will recreate arrow files anew each time
  logFile = createLogFile("onlyDay1")
)
```
```{r}
samples <- c(day5, day10, day25, day4dpi)
# # Set understandable names
names(samples) <- c("d05", "d10", "d25", "Inf-d4")
inputFiles <- c('scATAC' = samples)
print(names(inputFiles))

ArrowFiles <- createArrowFiles(
  inputFiles = inputFiles,
  sampleNames = names(samples),
  # filterFrags = 2500,
  minTSS = 5, # 4 was the default value. Dont set this too high;can always increase later
  minFrags = 2500,
  addTileMat = TRUE,
  addGeneScoreMat = TRUE,
  subThreading = TRUE,
  cleanTmp = TRUE,
  force = TRUE, # will recreate arrow files anew each time
  logFile = createLogFile("All5")
)
```

## Creating Arrow Files: 

Each Arrow file stores all of the data associated with an individual sample (i.e. metadata, accessible fragments, and data matrices).
```{r}
# ArrowFiles <- createArrowFiles(
#   inputFiles = inputFiles,
#   sampleNames = names(inputFiles),
#   # filterFrags = 2500,
#   minTSS = 5, # 4 was the default value. Dont set this too high;can always increase later
#   minFrags = 2500,
#   addTileMat = TRUE,
#   addGeneScoreMat = TRUE,
#   subThreading = TRUE,
#   cleanTmp = TRUE,
#   force = FALSE, # will recreate arrow files anew each time
#   logFile = createLogFile("All5")
# )
```
# Quality Control

## Doublet Score

```{r}
doubScores <- addDoubletScores(
  input = c(ArrowFiles1, ArrowFiles),
  k = 30, # Refers to how many cells near a "pseudo-doublet" to count
  knnMethod = "LSI", # Refers to the embedding to use for nearest neighbor search.
  LSIMethod = 1,
  force = FALSE,
  verbose = TRUE
)
```

# ArchR Project

```{r}
ATACSeq_project_All5 <- ArchRProject(
  ArrowFiles = c(ArrowFiles1, ArrowFiles),
  outputDirectory = "/home/keshavprasad/Documents/scATACseq/All5/",
  copyArrows = FALSE # This is recommended so that you maintain an unaltered copy for later usage.
)
paste0("Memory Size = ", round(object.size(ATACSeq_project_All5) / 10^6, digits = 3), " MB")
getAvailableMatrices(ATACSeq_project_All5)

# Inspect the newly created Project.
print(ATACSeq_project_All5)
# Save the newly created
saveArchRProject(
  ArchRProj = ATACSeq_project_All5,
  outputDirectory = getOutputDirectory(ATACSeq_project_All5),
  overwrite = TRUE,
  load = TRUE,
  logFile = createLogFile("saveArchRProject_All5"),
)
```


# Plotting QC metrics 

### UniqueFrags Vs Enrichment

```{r}
QCMetric_df <- getCellColData(ATACSeq_project_All5, select = c("log10(nFrags)", "TSSEnrichment"))
```
```{r fig.align='center', fig.height=6, fig.width=6}
(scatterplot_UniqueFragsVsEnrichment_PreFiltering <- ggPoint(
  x = QCMetric_df[, 1],
  y = QCMetric_df[, 2],
  colorDensity = TRUE,
  continuousSet = "sambaNight",
  xlabel = "Log10 Unique Fragments",
  ylabel = "TSS Enrichment",
  xlim = c(log10(500), quantile(QCMetric_df[, 1], probs = 0.99)),
  ylim = c(0, quantile(QCMetric_df[, 2], probs = 0.99))
) +
  geom_hline(yintercept = 5, lty = "dashed") +
  geom_vline(xintercept = 3.4, lty = "dashed") +
  theme(axis.title = element_text(size = 15), axis.text = element_text(size = 15)))

plotPDF(scatterplot_UniqueFragsVsEnrichment_PreFiltering,
  name = "TSS-vs-Frags_PreFiltering.pdf",
  ArchRProj = ATACSeq_project_All5
)
```

```{r}
create_GroupPlot <- function(projectname, cellColData_colname) {
  Group_Plot <- plotGroups(
    ArchRProj = projectname,
    groupBy = "Sample",
    colorBy = "cellColData",
    name = cellColData_colname,
    plotAs = "violin",
    alpha = 0.4,
    addBoxPlot = TRUE
  )
  return(Group_Plot)
}

my_gg_theme <- theme(
  axis.title = element_text(size = 15),
  axis.text = element_text(size = 15),
  legend.text = element_text(size = 10)
) # To have more readable font and axis labels.
```
### violin plot for each sample for TSSEnrichment.
```{r fig.align='center', fig.height=7, fig.width=7}
(Group_plot_TSS_violin_PreFiltering <- create_GroupPlot(ATACSeq_project_All5, "TSSEnrichment") + ggtitle("TSSEnrichment")) + my_gg_theme
```
### Plots (per sample) for log10 (unique nuclear fragments) 
```{r fig.align='center', fig.height=7, fig.width=7}
(Group_plot_nFrags_violin_PreFiltering <- create_GroupPlot(ATACSeq_project_All5, "log10(nFrags)") + ggtitle("nFrags")) + my_gg_theme
```
### Plots (per sample) for BlacklistRatio
```{r fig.align='center', fig.height=7, fig.width=7}
(Group_plot_BLR_violin_PreFiltering <- create_GroupPlot(ATACSeq_project_All5, "BlacklistRatio") + ggtitle("BlacklistRatio")) + my_gg_theme
```
### Plots (per sample) for NucleosomeRatio
```{r fig.align='center', fig.height=7, fig.width=7}
(Group_plot_NR_violin_PreFiltering <- create_GroupPlot(ATACSeq_project_All5, "NucleosomeRatio") + ggtitle("NucleosomeRatio") + my_gg_theme )
```
### Plots (per sample) for DoubletScore
```{r fig.align='center', fig.height=7, fig.width=7}
(Group_plot_DS_violin_PreFiltering <- create_GroupPlot(ATACSeq_project_All5, "DoubletScore") + ggtitle("DoubletScore")) + my_gg_theme
```
### Plots (per sample) for DoubletEnrichment
```{r fig.align='center', fig.height=7, fig.width=7}
(Group_plot_DE_violin_PreFiltering <- create_GroupPlot(ATACSeq_project_All5, "DoubletEnrichment") + ggtitle("DoubletEnrichment")) + my_gg_theme
```
```{r}
plotPDF(Group_plot_TSS_violin_PreFiltering, Group_plot_nFrags_violin_PreFiltering,
        Group_plot_BLR_violin_PreFiltering, Group_plot_NR_violin_PreFiltering,
        Group_plot_DS_violin_PreFiltering, Group_plot_DE_violin_PreFiltering, 
        name = "QC-Metrics_PreFiltering.pdf", 
        ArchRProj = ATACSeq_project_All5, 
        width = 12, height = 12)
```

```{r fig.align='center', fig.height=7, fig.width=7} 
(FragSizePlot <- plotFragmentSizes(ArchRProj = ATACSeq_project_All5)) + my_gg_theme
(TSSEnrichmentPlot <- plotTSSEnrichment(ArchRProj = ATACSeq_project_All5)) + my_gg_theme
plotPDF(FragSizePlot, TSSEnrichmentPlot, name = "QC-Sample-FragSizes-TSSProfile_PreFiltering.pdf", ArchRProj = ATACSeq_project_All5, width = 10, height = 10)
```
```{r}
## Finding doublets
ATACSeq_project_All5 <- filterDoublets(ArchRProj = ATACSeq_project_All5)
``` 
# Filtering 

Filter out low Quality Cells from ArchR Project, based on the violin plots of QC Metrics. 

```{r}
ATACSeq_project_All5 <- ATACSeq_project_All5[ATACSeq_project_All5$NucleosomeRatio < 2.5 &
  ATACSeq_project_All5$TSSEnrichment > 6 &
  ATACSeq_project_All5$BlacklistRatio < 0.05 &
  ATACSeq_project_All5$DoubletScore < 50 &
  ATACSeq_project_All5$DoubletEnrichment < 3.5]
```

# QC Plots PostFiltering

### violin plot for each sample for TSSEnrichment.
```{r fig.align='center', fig.height=7, fig.width=7}
(Group_plot_TSS_violin_PostFiltering <- create_GroupPlot(ATACSeq_project_All5, "TSSEnrichment") + ggtitle("TSSEnrichment")) + my_gg_theme
```
### Plots (per sample) for log10 (unique nuclear fragments) #### log10 (unique nuclear fragments)
```{r fig.align='center', fig.height=7, fig.width=7}
(Group_plot_nFrags_violin_PostFiltering <- create_GroupPlot(ATACSeq_project_All5, "log10(nFrags)") + ggtitle("nFrags")) + my_gg_theme
```

### Plots (per sample) for BlacklistRatio
```{r fig.align='center', fig.height=7, fig.width=7}
(Group_plot_BLR_violin_PostFiltering <- create_GroupPlot(ATACSeq_project_All5, "BlacklistRatio") + ggtitle("BlacklistRatio")) + my_gg_theme
```

### Plots (per sample) for NucleosomeRatio
```{r fig.align='center', fig.height=7, fig.width=7}
(Group_plot_NR_violin_PostFiltering <- create_GroupPlot(ATACSeq_project_All5, "NucleosomeRatio") + ggtitle("NucleosomeRatio")) + my_gg_theme
```

### Plots (per sample) for DoubletScore
```{r fig.align='center', fig.height=7, fig.width=7}
(Group_plot_DS_violin_PostFiltering <- create_GroupPlot(ATACSeq_project_All5, "DoubletScore") + ggtitle("DoubletScore")) + my_gg_theme
```

### Plots (per sample) for DoubletEnrichment
```{r fig.align='center', fig.height=7, fig.width=7}
(Group_plot_DE_violin_PostFiltering <- create_GroupPlot(ATACSeq_project_All5, "DoubletEnrichment") + ggtitle("DoubletEnrichment")) + my_gg_theme
```
```{r}
plotPDF(Group_plot_TSS_violin_PostFiltering, Group_plot_nFrags_violin_PostFiltering, Group_plot_BLR_violin_PostFiltering, Group_plot_NR_violin_PostFiltering, Group_plot_DS_violin_PostFiltering, Group_plot_DE_violin_PostFiltering, name = "QC-Metrics_PostFiltering.pdf", ArchRProj = ATACSeq_project_All5, width = 12, height = 12)
```

### Plotting Sample Fragment Size Distribution and TSS Enrichment Profiles.
```{r fig.align='center', fig.height=7, fig.width=7}
(FragSizePlot_PostFiltering <- plotFragmentSizes(ArchRProj = ATACSeq_project_All5)) + my_gg_theme
(TSSEnrichmentPlot_PostFiltering <- plotTSSEnrichment(ArchRProj = ATACSeq_project_All5)) + my_gg_theme
plotPDF(FragSizePlot_PostFiltering, TSSEnrichmentPlot_PostFiltering, name = "QC-Sample-FragSizes-TSSProfile_PostFiltering.pdf", ArchRProj = ATACSeq_project_All5, width = 10, height = 10)
```

```{r}
## Finding doublets
# ATACSeq_project_All5 <- filterDoublets(ArchRProj = ATACSeq_project_All5)
```

```{r}
saveArchRProject(
  ArchRProj = ATACSeq_project_All5,
  outputDirectory = getOutputDirectory(ATACSeq_project_All5),
  overwrite = TRUE,
  load = TRUE,
  logFile = createLogFile("saveArchRProject_All5"),
)
```

```{r}
ATACSeq_project_All5 <- loadArchRProject(path = "/home/keshavprasad/Documents/scATACseq/All5/",
  force = FALSE, showLogo = TRUE)

my_gg_theme <- theme(
  axis.title = element_text(size = 15),
  axis.text = element_text(size = 15),
  legend.text = element_text(size = 10)
)
```



# Dimensional Reduction, INTEGRATION, CLUSTERING & EMBEDDING

```{r}
ATACSeq_project_All5 <- addIterativeLSI( # 4.1 ## Reducing Dims via Iterative LSI.# The most common parameters to tweak are iterations, varFeatures, and resolution.
  ArchRProj = ATACSeq_project_All5,
  useMatrix = "TileMatrix",
  name = "IterativeLSI_all5",
  # iterations = 4,
  saveIterations = TRUE,
  clusterParams = list(resolution = c(0.2),
                       sampleCells = 10000,
                       n.start = 10),
  varFeatures = 15000,
  dimsToUse = 1:30,
  force = TRUE
)
```
Using LSI object as Input to correct for Batch Effects via Harmony
```{r}
ATACSeq_project_All5 <- addHarmony( # 4.4 ## Batch Effect Correction with HARMONY                   ##
  ArchRProj = ATACSeq_project_All5,
  reducedDims = "IterativeLSI_all5",
  name = "Harmony_all5",
  groupBy = "Sample",
  verbose = TRUE,
  force = TRUE
)
```
```{r}
ATACSeq_project_All5 <- addUMAP( # 5.1 ## Visualising clusters via UMAP
  ArchRProj = ATACSeq_project_All5,
  reducedDims = "Harmony_all5", # Need to use the Harmony object for clustering and UMAP for proper integration
  name = "UMAP_all5",
  nNeighbors = 30,
  minDist = 0.4,
  metric = "cosine",
  force = TRUE,
  verbose = TRUE,
  saveModel = TRUE
)
```

```{r fig.align='center', fig.height=7, fig.width=7}
# Create UMAP for all samples together
(p1_UMAP <- plotEmbedding(
  ArchRProj = ATACSeq_project_All5,
  colorBy = "cellColData",
  name = "Sample",
  embedding = "UMAP_all5",
  keepAxis = TRUE
) + my_gg_theme)
# plotPDF(p1_UMAP, name = "Plot-UMAP-Samplewise-Clusters.pdf", ArchRProj = ATACSeq_project_All5, , width = 10, height = 10)
```

## Create SplitView UMAP

```{r}
# Create a Split view UMAP so that the UMAP for each sample can be visualized separately.
create_splitUMAP <- function(ATACSeq_project_All5, sampleObject) {
  split_UMAP <- plotEmbedding(
    ArchRProj = ATACSeq_project_All5,
    colorBy = "cellColData",
    name = "Sample",
    embedding = "UMAP_all5",
    highlightCells = getCellNames(ArchRProj = ATACSeq_project_All5)[which(ATACSeq_project_All5@cellColData$Sample == sampleObject)],
    keepAxis = TRUE,
    baseSize = 10,
    plotAs = "points"
  )
  return(split_UMAP)
}
```
```{r fig.align='center', fig.height=7, fig.width=7} 
# Get the the Samples from the ArchR project
samplelist <- ATACSeq_project_All5@cellColData$Sample
# Call the function to create UMAP for each sample
(d1_UMAP <- create_splitUMAP(ATACSeq_project_All5, samplelist@values[1]) + my_gg_theme)
(d5_UMAP <- create_splitUMAP(ATACSeq_project_All5, samplelist@values[5]) + my_gg_theme)
(d10_UMAP <- create_splitUMAP(ATACSeq_project_All5, samplelist@values[4]) + my_gg_theme)
(d25_UMAP <- create_splitUMAP(ATACSeq_project_All5, samplelist@values[3]) + my_gg_theme)
(d4pi_UMAP <- create_splitUMAP(ATACSeq_project_All5, samplelist@values[2]) + my_gg_theme)

# (ggAlignPlots(p1_UMAP, p2_UMAP, type = "h"))
plotPDF(d1_UMAP, d5_UMAP, d10_UMAP, d25_UMAP, d4pi_UMAP,
  name = "Plot-Splitview UMAP-Sample-Clusters.pdf",
  ArchRProj = ATACSeq_project_All5, 
  width = 10,
  height = 10
)
```
# Adding Clusters
```{r fig.align='center', fig.height=7, fig.width=7}
# Adding Clusters
ATACSeq_project_All5 <- addClusters( # 5.1 ## Creating clusters based on Iterative LSI object      ##
  input = ATACSeq_project_All5,
  reducedDims = "Harmony_all5", # Harmony_all5
  method = "Seurat",
  name = "Clusters_all5",
  resolution = 0.5,
  force = TRUE
)
ATACSeq_project_All5 <- addUMAP( # 5.1 ## Visualising clusters via UMAP
  ArchRProj = ATACSeq_project_All5,
  reducedDims = "Harmony_all5", # Need to used the HArmony object for cluistering and UMAP for proper integration
  name = "ClustersUMAP_all5",
  nNeighbors = 30,
  minDist = 0.4,
  metric = "cosine",
  force = TRUE,
  verbose = TRUE,
  saveModel = TRUE
)

(p2_UMAP <- plotEmbedding(ArchRProj = ATACSeq_project_All5, colorBy = "cellColData", name = "Clusters_all5", embedding = "ClustersUMAP_all5", keepAxis = TRUE) + my_gg_theme)
plotPDF(p2_UMAP, name = "Plot-UMAP-Clusters.pdf", ArchRProj = ATACSeq_project_All5, , width = 10, height = 10)

# # Accessing the clusters
head(ATACSeq_project_All5$Clusters_all5)
# tabulate the number of cells present in each cluster:
table(ATACSeq_project_All5$Clusters_all5)

# To better understand which samples reside in which clusters, we can create a
# cluster confusion matrix across each sample using the confusionMatrix() function.
cM <- confusionMatrix(
  paste0(ATACSeq_project_All5$Clusters_all5),
  paste0(ATACSeq_project_All5$Sample)
)
print(cM)

# plotting the Confusion Matrix
cM <- cM / Matrix::rowSums(cM)
(p <- pheatmap::pheatmap(
  mat = as.matrix(cM), color = paletteContinuous("whiteBlue"),
  border_color = "black"
) + my_gg_theme)
```

#  Gene Scores and Marker Genes
## Identifying Marker Genes
```{r}
markersGS <- getMarkerFeatures(
  ArchRProj = ATACSeq_project_All5,
  useMatrix = "GeneScoreMatrix",
  groupBy = "Clusters_all5",
  bias = c("TSSEnrichment", "log10(nFrags)"),
  testMethod = "wilcoxon"
)

## Markers List
markerList <- getMarkers(markersGS, cutOff = "FDR <= 0.05 & Log2FC >= 1.0")
# Create an empty list to store dataframes
df_list <- list()

# Loop over the list and convert each element to a dataframe
for (i in 1:length(markerList)) {
  cluster_name <- names(markerList)[i]
  cluster_df <- as.data.frame(markerList[[i]])

  # Add the dataframe to the list with a unique name
  df_list[[paste0(cluster_name, "_sheet")]] <- cluster_df
}

# Write the dataframes to different sheets of an Excel file
write_xlsx(df_list, path = "Clusters_res0.1_ATACseq.xlsx")
```
# Marker Genes
##  visualize all of the marker features simultaneously
```{r}
markerGenesList <- c("Plag2g2a", "Defa-rs1", "Mmp7", "Lyz1", "Spdef", "Tcf7l2", "Ephb3", "Sis", "Ada", "Lct")

(heatmapGS <- plotMarkerHeatmap(
  seMarker = markersGS,
  cutOff = "FDR <= 0.01 & Log2FC >= 1.0",
  limits = c(-3, 3),
  returnMatrix = FALSE,
  plotLog2FC = TRUE,
  labelMarkers = markerGenesList,
  transpose = TRUE,
  labelRows = TRUE, clusterCols = TRUE, nPrint = 10
))

plotPDF(heatmapGS,
  name = "GeneScores-Marker-Heatmap", width = 10, height = 10,
  ArchRProj = ATACSeq_project_All5,
)
```

# Annotating Cell types with a Reference Dataset 

```{r fig.align='center', fig.height=7, fig.width=7}
# Read-in the reference
annotate_reference <- readRDS(file.path("/media/keshavprasad/HornefLab_Data3/scRNA_AnnotationData_Johannes", "scrna_with_day25.Rds"))

# add gene integration matrix
ATACSeq_project_All5 <- addGeneIntegrationMatrix(
  ArchRProj = ATACSeq_project_All5,
  useMatrix = "GeneScoreMatrix",
  matrixName = "GeneIntegrationMatrix",
  reducedDims = "IterativeLSI_all5",
  seRNA = annotate_reference,
  addToArrow = FALSE,
  groupRNA = "int_0.3_broad_tuft",
  nameCell = "predictedCell_Un",
  nameGroup = "predictedGroup_Un",
  nameScore = "predictedScore_Un"
)

# Plot UMAP with predicted cell types
(annotated_UMAP_scRNAseqRef <- plotEmbedding(ATACSeq_project_All5,
  name = "predictedGroup_Un",
  embedding = "UMAP_all5",
  size = 1.5,
  labelAsFactors = F,
  labelMeans = F
) + my_gg_theme)

plotPDF(annotated_UMAP_scRNAseqRef,
  name = "annotated_UMAP_scRNAseqRef_unconstrained",
  width = 10, height = 10, ArchRProj = ATACSeq_project_All5,
)
```
